sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c

```diff
@@ -720,25 +720,14 @@ static void _receive(gnrc_pktsnip_t *pkt)
         iface = ((gnrc_netif_hdr_t *)netif->data)->if_pid;
     }
 
-    if ((pkt->next != NULL) && (pkt->next->type == GNRC_NETTYPE_IPV6) &&
-        (pkt->next->size == sizeof(ipv6_hdr_t))) {
-        /* IP header was already marked. Take it. */
-        ipv6 = pkt->next;
-
-        if (!ipv6_hdr_is(ipv6->data)) {
-            DEBUG("ipv6: Received packet was not IPv6, dropping packet\n");
-            gnrc_pktbuf_release(pkt);
-            return;
-        }
-#ifdef MODULE_GNRC_IPV6_WHITELIST
-        if (!gnrc_ipv6_whitelisted(&((ipv6_hdr_t *)(ipv6->data))->src)) {
-            DEBUG("ipv6: Source address not whitelisted, dropping packet\n");
-            gnrc_pktbuf_release(pkt);
-            return;
+    for (ipv6 = pkt; ipv6 != NULL; ipv6 = ipv6->next) { /* find IPv6 header if already marked */
```

Comment:
I still don't get it. Either the first snip is marked wrongly as IPv6 (shouldn't we rather raise an assert for that?) or there are several valid IPv6 headers in which case we would simply find the first?